ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala

```diff
@@ -0,0 +1,439 @@
+              logger.trace(s"Validating a batch of ${batch.getSubmissionsCount} submissions")
+              Metrics.batchSizes.update(batch.getSubmissionsCount)
+              Metrics.receivedBatchSubmissionBytes.update(batch.getSerializedSize)
+              processBatch(
+                participantId,
+                correlationId,
+                recordTime,
+                batchSubmissionSource(batch, correlationId),
+                ledgerStateReader,
+                commitStrategy)
+
+            case Right(other) =>
+              Future.failed(
+                ValidationFailed.ValidationError(
+                  s"Unexpected message in envelope: ${other.getClass.getSimpleName}"))
+
+            case Left(error) =>
+              Future.failed(ValidationFailed.ValidationError(s"Cannot open envelope: $error"))
+          }
+
+          result
+            .andThen {
+              case Failure(exception) =>
+                logger.error(s"Validation failure: $exception")
+              case Success(_) =>
+                ()
+            }
+        }
+      )
+    }
+
+  private def singleSubmissionSource(
+      envelope: ByteString,
+      submission: DamlSubmission,
+      correlationId: CorrelationId): Source[Stage1, NotUsed] = {
+    val logEntryId = bytesToLogEntryId(envelope)
+    Source.single(Indexed(CorrelatedSubmission(correlationId, logEntryId, submission), 0L))
+  }
+
+  private def batchSubmissionSource(batch: DamlSubmissionBatch, correlationId: CorrelationId)(
+      implicit executionContext: ExecutionContext): Source[Stage1, NotUsed] =
+    Source(
+      Indexed
+        .fromSeq(batch.getSubmissionsList.asScala
+          .map(cs => cs.getCorrelationId -> cs.getSubmission))
+        .to)
+      .mapAsyncUnordered(params.cpuParallelism) {
+        _.mapFuture {
+          case (correlationId, submissionEnvelope) =>
+            // Decompress and decode the submissions in parallel.
+            Timed.timedAndTrackedFuture(
+              Metrics.decode,
+              Metrics.decodeRunning,
+              Future {
+                val submission = Envelope
+                  .openSubmission(submissionEnvelope)
+                  .fold(error => throw validator.ValidationFailed.ValidationError(error), identity)
+                Metrics.receivedSubmissionBytes.update(submission.getSerializedSize)
+                CorrelatedSubmission(
+                  correlationId,
+                  bytesToLogEntryId(submissionEnvelope),
+                  submission)
+              }
+            )
+        }
+      }
+
+  private type DamlInputState = Map[DamlStateKey, Option[DamlStateValue]]
+
+  //
+  // The following type aliases describe how the batch processing pipeline transforms the data:
+  //
+
+  // The batch pipeline starts with a stream of correlated submissions that carry their original index
+  // in the batch.
+  private type Stage1 = Indexed[CorrelatedSubmission]
```

Comment:
I find the naming here confusing. I would have expected a "stage" to be a function that goes from `A` to `B`. I think these might be better named as "results" or "outputs" (with the first one being the "input").

Reply:
Renamed the types to be `Inputs` and `Outputs1`, `Outputs2`, etc. Can you check if this is less confusing, please?

Reply:
Sounds good to me.